File: modules/service_recon.py
# This Python file uses the following encoding: utf-8

# MODULES AND/OR LIBRARIES
from socket import getservbyport, error as socket_error
from logging import debug, info, error
from threading import Lock
from csv import DictReader
from json import load as jsonload

##############################

# GLOBAL VARIABLES

##############################

_services = {}
_services_loaded = False
_services_lock = Lock()
_probes = None

##############################

# LOAD IANA SERVICE-NAMES CSV FILE

##############################

def load_iana_services(path="data/service-names-port-numbers.csv"):
    global _services, _services_loaded
    with _services_lock:
        if _services_loaded:
            return

        try:
            with open(path, "r", encoding="utf-8") as file:
                reader = DictReader(file)
                for row in reader:
                    service_name = row["Service Name"].strip()
                    port = row["Port Number"].strip()
                    protocol = row["Transport Protocol"].strip().lower()

                    if not service_name or not port or not protocol:
                        continue

                    try:
                        port_number = int(port)
                        _services[(port_number, protocol)] = service_name
                    except ValueError:
                        continue

            _services_loaded = True
            debug(f"Loaded services from {path}")

        except Exception as e:
            error(f"Failed to load IANA services: {e}")

##############################

# SERVICE NAME LOOKUP

##############################

def lookup_service_name(port, protocol="tcp"):
    protocol = protocol.lower()

    # 1. Try system-level getservbyport
    try:
        service = getservbyport(port, protocol)
        debug(f"Resolved service via getservbyport: {service} on port {port}/{protocol}")
        return service
    except OSError:
        pass

    # 2. Try loaded IANA service names
    load_iana_services()
    service = _services.get((port, protocol), "unknown")
    debug(f"Resolved service via IANA services: {service} on port {port}/{protocol}")
    return service

##############################

# LOAD SERVICE PROBES

##############################

def load_probes(path="data/service-probes.json"):
    global _probes
    if _probes is not None:
        return _probes
    with open(path, "r", encoding="utf-8") as f:
        _probes = jsonload(f)
    # Decode escape sequences in payloads
    for probe in _probes:
        probe["probe_payload"] = probe["probe_payload"].encode("utf-8").decode("unicode_escape")
        debug(probe)
    return _probes

##############################

# FIND SERVICE PROBES

##############################

def find_probe(port, protocol):
    probes = load_probes()
    protocol = protocol.lower()
    for probe in probes:
        if probe["port"] == port and probe["protocol"] == protocol:
            return probe
    return None

##############################

# BANNER GRABBER

##############################

def grab_service_banner(sock, timeout=2, port=None, protocol="tcp"):
    banner = ""
    try:
        sock.settimeout(timeout)
        probe = None

        # For HTTP(S) ports, send explicit GET request to get headers only
        if port in (80, 443) and protocol.lower() == "tcp":
            probe = "GET / HTTP/1.0\r\n\r\n"
            debug(f"Sending HTTP GET probe for port {port}")

        else:
            if port is not None:
                probe_data = find_probe(port, protocol)
                if probe_data:
                    probe = probe_data["probe_payload"]
                    debug(f"Sending probe payload for {protocol.upper()} port {port}: {probe.strip()}")

        if probe:
            sock.sendall(probe.encode())

        data_chunks = []
        while True:
            try:
                data = sock.recv(4096)
                if not data:
                    break
                data_chunks.append(data)
            except Exception:
                break

        if data_chunks:
            full_banner = b"".join(data_chunks).decode(errors='ignore').strip()
            debug(f"Received full banner:\n{full_banner}")

            # For HTTP, extract headers (everything before first double CRLF)
            if port in (80, 443) and protocol.lower() == "tcp":
                header_end = full_banner.find("\r\n\r\n")
                if header_end != -1:
                    banner = full_banner[:header_end].strip()
                else:
                    banner = full_banner  # fallback: entire banner if no header boundary found
            else:
                banner = full_banner

        info("Banner grabbing completed.")

    except Exception as e:
        error(f"Failed to grab banner: {e}")

    return banner

File: modules/service_recon.go
package modules

import (
	"encoding/csv"
	"encoding/json"
	"fmt"
	"net"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

// ##############################
//
// # GLOBAL VARIABLES
//
// ##############################

var (
	servicePortMap = make(map[string]string) // key: "port/protocol"
	serviceOnce    sync.Once

	serviceProbes []Probe
	probesOnce    sync.Once
	probesErr     error
)

// Probe represents a service probe structure
type Probe struct {
	Port         int    `json:"port"`
	Protocol     string `json:"protocol"`
	ProbePayload string `json:"probe_payload"`
}

// ##############################
//
// # LOAD IANA SERVICE-NAMES CSV FILE
//
// ##############################

func LoadIanaServices(filePath string) error {
	var loadErr error

	serviceOnce.Do(func() {
		file, err := os.Open(filePath)
		if err != nil {
			loadErr = err
			return
		}
		defer file.Close()

		reader := csv.NewReader(file)
		headers, err := reader.Read()
		if err != nil {
			loadErr = err
			return
		}

		colIndex := make(map[string]int)
		for i, col := range headers {
			colIndex[strings.TrimSpace(col)] = i
		}

		for {
			record, err := reader.Read()
			if err != nil {
				break // EOF
			}

			serviceName := strings.TrimSpace(record[colIndex["Service Name"]])
			portStr := strings.TrimSpace(record[colIndex["Port Number"]])
			protocol := strings.ToLower(strings.TrimSpace(record[colIndex["Transport Protocol"]]))

			if serviceName == "" || portStr == "" || protocol == "" {
				continue
			}

			port, err := strconv.Atoi(portStr)
			if err != nil {
				continue
			}

			key := fmt.Sprintf("%d/%s", port, protocol)
			servicePortMap[key] = serviceName
		}
	})

	return loadErr
}

// ##############################
//
// # SERVICE NAME LOOKUP
//
// ##############################

func LookupServiceName(port int, protocol string) string {
	key := fmt.Sprintf("%d/%s", port, strings.ToLower(protocol))
	if service, exists := servicePortMap[key]; exists {
		return service
	}
	return "unknown"
}

// ##############################
//
// # LOAD SERVICE PROBES (JSON)
//
// ##############################

func LoadProbes(path string) ([]Probe, error) {
	probesOnce.Do(func() {
		data, err := os.ReadFile(path)
		if err != nil {
			probesErr = err
			return
		}

		err = json.Unmarshal(data, &serviceProbes)
		if err != nil {
			probesErr = err
			return
		}
	})
	return serviceProbes, probesErr
}

// ##############################
//
// # FIND SERVICE PROBE
//
// ##############################

func FindProbe(port int, protocol string) *Probe {
	probes, err := LoadProbes("data/service-probes.json")
	if err != nil {
		return nil
	}
	for _, probe := range probes {
		if probe.Port == port && strings.EqualFold(probe.Protocol, protocol) {
			return &probe
		}
	}
	return nil
}

// ##############################
//
// # DECODE PROBE PAYLOAD
//
// ##############################

func decodeProbePayload(payload string) ([]byte, error) {
	decodedStr, err := strconv.Unquote(`"` + payload + `"`)
	if err != nil {
		return nil, err
	}
	return []byte(decodedStr), nil
}

// ##############################
//
// # BANNER GRABBER
//
// ##############################

func GrabServiceBanner(conn net.Conn, timeout time.Duration, port int, protocol string) string {
	var banner strings.Builder

	_ = conn.SetDeadline(time.Now().Add(timeout))
	defer conn.SetDeadline(time.Time{})

	// Send optional probe payload
	if probe := FindProbe(port, protocol); probe != nil {
		payloadBytes, err := decodeProbePayload(probe.ProbePayload)
		if err == nil && len(payloadBytes) > 0 {
			_, _ = conn.Write(payloadBytes)
		}
	}

	buf := make([]byte, 4096)
	for {
		n, err := conn.Read(buf)
		if n > 0 {
			banner.Write(buf[:n])
		}
		if err != nil {
			break // EOF or timeout
		}
	}

	fullBanner := banner.String()

	// Extract HTTP headers if possible (everything before first double CRLF)
	headersEnd := strings.Index(fullBanner, "\r\n\r\n")
	if headersEnd != -1 {
		return strings.TrimSpace(fullBanner[:headersEnd])
	}

	// Fallback: return entire response if no header boundary found
	return strings.TrimSpace(fullBanner)
}